build: refactor Makefile invocations to support parallel make (-j)

X-sam opened the pull request:
<!--
Thanks for taking the time to open a pull request! Please make sure you've read the "Opening Pull Requests" section of our Contributing Guide:

https://github.com/Opentrons/opentrons/blob/edge/CONTRIBUTING.md#opening-pull-requests

To ensure your code is reviewed quickly and thoroughly, please fill out the sections below to the best of your ability!
-->

# Overview
This PR changes how our python Makefile targets work so that, instead of being managed linearly, they're handled as peers, allowing us to Make python targets in parallel (with the `-j` flag)
<!--
Use this section to describe your pull-request at a high level. If the PR addresses any open issues, please tag the issues here.
-->

# Changelog
 - Added `teardown` target to `update-server` Makefile, for consistency.
 - Refactored `setup-py` target
 - Refactored `clean-py` target
 - Refactored `teardown-py` target
 -  Refactored `lint-py` target
 
<!--
List out the changes to the code in this PR. Please try your best to categorize your changes and describe what has changed and why.

Example changelog:
- Fixed app crash when trying to calibrate an illegal pipette
- Added state to API to track pipette usage
- Updated API docs to mention only two pipettes are supported

IMPORTANT: MAKE SURE ANY BREAKING CHANGES ARE PROPERLY COMMUNICATED
-->

# Review requests
Check all the python make flows, `make clean-py` `make setup-py` `make teardown-py` `make lint-py`. Try invoking with varying numbers on -j: `make -j 2 clean-py` etc. Ensure your workflow works correctly
<!--
Describe any requests for your reviewers here.
-->

# Risk assessment
This affects our python make flows, which affects any functionality that relies on making python parts. As the refactored functionality should be at feature-parity, this should have minimal risk- though we should keep an eye open for bugs in this new implementation.

<!--
Carefully go over your pull request and look at the other parts of the codebase it may affect. Look for the possibility, even if you think it's small, that your change may affect some other part of the system - for instance, changing return tip behavior in protocol may also change the behavior of labware calibration.

Identify the other parts of the system your codebase may affect, so that in addition to your own review and testing, other people who may not have the system internalized as much as you can focus their attention and testing there.
-->

Commit:
build: refactor Makefile invocations to support parallel make (-j)

This changes how our python setup/lint/teardown/clean work so that, instead of being managed
linearly as a recipe, they're handled as dependencies of the rule which properly maps their
relationship (e.g. neighbors, not dependents)

mcous commented on Makefile:
```diff
@@ -47,44 +47,53 @@ usb_host=$(shell yarn run -s discovery find -i 169.254)
 
-.PHONY: setup-py
-setup-py:
+PYTHON_DIRS = $(API_DIR) $(UPDATE_SERVER_DIR) $(NOTIFY_SERVER_DIR) $(ROBOT_SERVER_DIR) $(SHARED_DATA_DIR)
```
Does this become easier if this list includes `shared-data/python` rather than `shared-data`? I think so, because `shared-data/python/Makefile` might (or could be changed to) follow the conventions in all the other Python makefiles, so then I'm not sure we'd need `PYTHON_TEARDOWN_DIRS` nor `PYTHON_LINT_DIRS`

X-sam replied:
That's a good idea- I'll give it a go!

X-sam replied:
26fa212 should address this- let me know what you think :)

mcous commented on Makefile:
```diff
@@ -47,44 +47,53 @@ usb_host=$(shell yarn run -s discovery find -i 169.254)
 
-.PHONY: setup-py
-setup-py:
+PYTHON_DIRS = $(API_DIR) $(UPDATE_SERVER_DIR) $(NOTIFY_SERVER_DIR) $(ROBOT_SERVER_DIR) $(SHARED_DATA_DIR)
```
Should this be ` :=` rather than `=`?

X-sam replied:
I suppose, though the way we're using it wouldn't have much difference

mcous commented on Makefile:
```diff
@@ -47,44 +47,53 @@ usb_host=$(shell yarn run -s discovery find -i 169.254)
-.PHONY: setup-py
-setup-py:
+PYTHON_DIRS = $(API_DIR) $(UPDATE_SERVER_DIR) $(NOTIFY_SERVER_DIR) $(ROBOT_SERVER_DIR) $(SHARED_DATA_DIR)
+PYTHON_CLEAN = $(addsuffix .clean, $(PYTHON_DIRS))
```
I'm unfamiliar with the convention (and most Makefile stuff, TBH...). Do you happen to have any good resources (or sample makefiles) to explain this style?

Also, are these targets no longer marked as `.PHONY`? Is that important for us?

X-sam replied:
Good catch, need a phony. Happy to talk more about the machinations of makefiles offline, if you'd like!

mcous replied:
I think I just got hung up on `.clean` vs something that looks a little more like our existing targets, e.g. `-clean`

X-sam replied:
changed this in 26fa212

X-sam replied:
`.PHONY` doesn't apply in the new pattern rule paradigm

mcous reviewed (commented):
Would you mind filling the PR template out with details from (HTML comments in) the PR template? In terms of "Review Requests", for this sort of build PR, **a test plan is really important** so that any reviewers can run through whatever `make` tasks on their machines and verify that it all works everywhere

X-sam commented:
> Would you mind filling the PR template out with details from (HTML comments in) the PR template? In terms of "Review Requests", for this sort of build PR, **a test plan is really important** so that any reviewers can run through whatever `make` tasks on their machines and verify that it all works everywhere

Yeah, of course! This should've been marked as draft RFC, apologies

SyntaxColoring commented:
What's the scope of this PR? Are you just looking to make our Python targets more parallel-safe, or fix parallelism in our makefiles in general?

If the latter, we still still have some hazards after these changes. Off the top of my head:

https://github.com/Opentrons/opentrons/blob/11c8328134998a117698085e58f1e891008c583b/shared-data/Makefile#L16

X-sam commented:
> What's the scope of this PR? Are you just looking to make our Python targets more parallel-safe, or fix parallelism in our makefiles in general?
> 
> If the latter, we still still have some hazards after these changes. Off the top of my head:
> 
> https://github.com/Opentrons/opentrons/blob/11c8328134998a117698085e58f1e891008c583b/shared-data/Makefile#L16

Specifically focused on the `setup` target, which I noticed was a slog to get through and easily addressed.

Commit:
build: added teardown target to update-server makefile

Commit:
build: simplify python targets

Feedback on PR was used to simplify our python setup, clean, teardown, and lint targets.

mcous reviewed (commented):
So I'm not 100% sure what's going on here, but it's definitely reproducible on my machine. For me, `make setup-py` and `make teardown-py` work just fine in non-parallel mode. However, in parallel mode (`-j`), I'm left with bad file permissions that cause future `setup-py` and `teardown-py` invocations (parallel or not) to fail.

Shot in the dark: this might have to do with pipenv installing some projects as editable dependencies of others, e.g. `shared-data/python` is a dependency of `api`, so perhaps order matters.

Example error during `make setup-py`:

```
PermissionError: [Errno 13] Permission denied: '/Volumes/Opentrons/opentrons/api/build/bdist.macosx-10.15-x86_64/wheel/opentrons'
gmake[2]: *** [Makefile:73: setup] Error 1
```

System details:

- macOS v10.15.7
- GNU Make v4.3
- Filesystem: case-sensitive APFS volume

X-sam commented:
> So I'm not 100% sure what's going on here, but it's definitely reproducible on my machine. For me, `make setup-py` and `make teardown-py` work just fine in non-parallel mode. However, in parallel mode (`-j`), I'm left with bad file permissions that cause future `setup-py` and `teardown-py` invocations (parallel or not) to fail.
> 
> Shot in the dark: this might have to do with pipenv installing some projects as editable dependencies of others, e.g. `shared-data/python` is a dependency of `api`, so perhaps order matters.
> 
> Example error during `make setup-py`:
> 
> ```
> PermissionError: [Errno 13] Permission denied: '/Volumes/Opentrons/opentrons/api/build/bdist.macosx-10.15-x86_64/wheel/opentrons'
> gmake[2]: *** [Makefile:73: setup] Error 1
> ```
> 
> System details:
> 
> * macOS v10.15.7
> * GNU Make v4.3

Interesting. Honestly, this terrified me for a minute because if that were true we'd be good and stuck. I *don't* believe it's the case, as the previous order involved building shared-data LAST! Will dig deeper though, pipenv is a nightmare world we can only hope to escape from soon.

mcous commented:
I've also got some build artifacts in places I didn't expect. Perhaps I got an upgraded version of setuptools installed or something during the setups and teardowns that's causing this issue? I must admit that the Python build ecosystem is a bit of a mystery to me

sfoster1 reviewed (approved):
Looks good to me, these workflows work with j2

mcous committed:
Merge branch 'edge' into make-cleanup

mcous committed:
fixup: make python clean a per-project-teardown-target dep

mcous committed:
fixup: consistent variable names
